File: cpupercentServer.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <array>
#include <csignal>
#include "percentgraph/percentgraphServer.hpp"

void sigHandler(int signal);
void setupSignals();


using server_t = PercentGraphServer<unsigned long, (16 + 1) * 2, 16, 1>;
server_t * server_ptr = nullptr;


int main(const int argv, const char * argc[]){

    server_t server("/home/dfuehrer/.cache/dwmblocks/cpupercent/cpupercent");
    server_ptr = &server;
    setupSignals();

    server.runServer([&server] (){
            std::ifstream statFile("/proc/stat");
            std::string cpu;
            std::array<server_t::stored_t, (server_t::numGraphPercents + server_t::numPrintPercents) * 2> saves;
            std::array<server_t::percent_t, server_t::numGraphPercents> percents;

            server_t::stored_t  user,   nice,   system,   idle,   iowait,   irq,   softirq,   steal,   tmp1,   tmp2;
            //statFile >> cpu >>  user >> nice >> system >> idle >> iowait >> irq >> softirq >> steal >> tmp1 >> tmp2;

            //server_t::percent_t active = user + system + nice + softirq + steal;
            //server_t::percent_t total  = active + idle + iowait;

            //saves = server.getDatas();

            //server_t::percent_t percent;
            //if(total == saves[1]){
            //    percent = 0;
            //}else{
            //    percent = ((active - saves[0]) * 100) / (total - saves[1]);
            //}
            //saves[0] = active;
            //saves[1] = total;

            saves = server.getDatas();

            server_t::percent_t percent;
            for(int i = 0; i < server_t::numGraphPercents + server_t::numPrintPercents; ++i){
            //for(int i = 0; i < 1; ++i){
                statFile >> cpu >>  user >> nice >> system >> idle >> iowait >> irq >> softirq >> steal >> tmp1 >> tmp2;

                server_t::stored_t active = user + system + nice + softirq + steal;
                server_t::stored_t total  = active + idle + iowait;

                server_t::percent_t tmpPercent;
                if(total == saves[i * 2 + 1]){
                    tmpPercent = 0;
                }else{
                    tmpPercent = ((active - saves[i * 2]) * 100) / (total - saves[i * 2 + 1]);
                }
                //std::cerr << cpu << ": " << tmpPercent << ", " << active - saves[i * 2] << " " << total - saves[i * 2 + 1] << std::endl;
                if(i){
                    percents[i - 1] = tmpPercent;
                }else{
                    percent = tmpPercent;
                }
                saves[i * 2]     = active;
                saves[i * 2 + 1] = total;

            }

            statFile.close();

            server.setPercents(percents, percent);
            server.saveDatas(saves);

    });

    server_ptr = nullptr;

    return 0;
}

void sigHandler(int signal){
    if(server_ptr != nullptr){
        server_ptr->stopRunning();
        server_ptr->cleanup();
    }
    // if this didnt work to kill things then it will at least remove the handler so next time the kill will succeed (ungracefully)
    std::signal(signal, SIG_DFL);
}


// set up handling for these normal signals, others will kill ungracefully
void setupSignals(){
    std::signal(SIGHUP,  sigHandler);
    std::signal(SIGTERM, sigHandler);
    std::signal(SIGINT,  sigHandler);
}

File: src/cpupercentServer.zig
const std = @import("std");
const percentgraph = @import("percentgraph");
//const percentgraph = @import("percentgraphServer.zig");
const params = @import("compile_params");

const num_percs = 1;
const server_t = percentgraph.PercentGraphServer(u64, (params.num_cpus + num_percs) * 2, params.num_cpus, num_percs);
var g_server: ?*server_t = null;

pub fn main() !void {
    setupSignals();
    var buf: [128]u8 = undefined;
    var fba = std.heap.FixedBufferAllocator.init(&buf);
    //defer fba.deinit();
    const alloc = fba.allocator();
    var cachedir_al = try percentgraph.getCacheDir(alloc, "dwmblocks/cpupercent");
    defer cachedir_al.deinit(alloc);
    try cachedir_al.appendSlice(alloc, "/cpupercent");
    var server = try server_t.init(cachedir_al.items);
    defer server.cleanup() catch unreachable;
    server.saveDatas(&[_]server_t.stored_t{0}**server_t.num_saved);
    g_server = &server;
    try server.runServerFunc(runServer);
    g_server = null;
}

fn runServer(server: *server_t) server_t.ClientRequestErrors!void {
    var statBuf: [1<<12]u8 = undefined;
    const statfile = std.fs.openFileAbsolute("/proc/stat", .{ .mode = .read_only }) catch return error.FileError;
    defer statfile.close();
    var statreader = statfile.reader(&statBuf);
    const saved = server.getDatas();
    var tosave: [server_t.num_saved]server_t.stored_t = undefined;
    var percent: [server_t.num_print_perc]server_t.stored_t = undefined;
    var cpu_percents: [server_t.num_graph_perc]server_t.percent_t = undefined;

    //var buf: [100]u8 = undefined;
    // user nice system idle iowait irq softirq steal guest guest_nice
    const active_inds = [_]bool{ true, true, true, false, false, true, true, true, true, true };
    const total_inds = [_]bool{ true, true, true, true, true, true, true, true, false, false };
    var cpu_line: usize = 0;
    while (cpu_line < server_t.num_graph_perc + server_t.num_print_perc) : (cpu_line += 1) {
        const line = statreader.interface.takeDelimiterExclusive('\n') catch return error.ReadError;
        const cpu_col = "cpu";
        if (!std.mem.eql(u8, line[0..cpu_col.len], cpu_col)) {
            // TODO 0 out data
            break;
        }

        var active: server_t.stored_t = 0;
        var total: server_t.stored_t = 0;
        var colnum: usize = 0;
        var ind = cpu_col.len;
        while (colnum < total_inds.len) : (colnum += 1) {
            while (line[ind] == ' ') {
                ind += 1;
            }
            const spaceind = ind + (std.mem.indexOfScalar(u8, line[ind..], ' ') orelse break);
            const colstr = line[ind..spaceind];
            const usage = std.fmt.parseUnsigned(server_t.stored_t, colstr, 10) catch return error.ReadError;
            ind = spaceind + 1;
            if (total_inds[colnum]) {
                total += usage;
                if (active_inds[colnum]) {
                    active += usage;
                }
            }
        }
        const perc = if ((total <= saved[cpu_line * 2 + 1]) or (active < saved[cpu_line * 2]))
            0
        else
            ((active - saved[cpu_line * 2]) * 100) / (total - saved[cpu_line * 2 + 1]);
        //var perc: server_t.stored_t = 0;
        //if ((total <= saved[cpu_line * 2 + 1]) or (active < saved[cpu_line * 2])) {
        //    std.debug.print("prev active: {}, total: {}; curr active: {}, total: {} ('{s}' -> '{s}')\n", .{ saved[cpu_line * 2], saved[cpu_line * 2 + 1], active, total, line });
        //} else {
        //    perc = ((active - saved[cpu_line * 2]) * 100) / (total - saved[cpu_line * 2 + 1]);
        //}
        if (cpu_line >= server_t.num_print_perc) {
            cpu_percents[cpu_line - server_t.num_print_perc] = @truncate(perc);
        } else {
            percent[cpu_line] = perc;
        }
        tosave[cpu_line * 2] = active;
        tosave[cpu_line * 2 + 1] = total;
    }
    //std.debug.print("percent overall: {}%, indiv percents: {any}\n", .{ percent[0], cpu_percents });

    server.setPercents(cpu_percents, &percent);
    server.saveDatas(&tosave);
}

fn sigHandler(signal: c_int) align(1) callconv(.c) void {
    // TODO figure out a way to do this other than a global var
    if (g_server) |server| {
        server.stopRunning();
        //server.cleanup() catch unreachable;
    }
    const dfl = std.posix.Sigaction{
        .handler = .{ .handler = std.posix.SIG.DFL },
        .mask = std.posix.sigemptyset(),
        .flags = 0,
    };
    std.posix.sigaction(@intCast(signal), &dfl, null);
}

fn setupSignals() void {
    const act = std.posix.Sigaction{
        .handler = .{ .handler = sigHandler },
        .mask = std.posix.sigemptyset(),
        .flags = 0,
    };
    std.posix.sigaction(std.posix.SIG.HUP, &act, null);
    std.posix.sigaction(std.posix.SIG.TERM, &act, null);
    std.posix.sigaction(std.posix.SIG.INT, &act, null);
}
